### ISSUE TITLE
Large Word2Vec model deserialization fails with both jblas and jcublas

### ISSUE THREAD

**Issue by danielchalef** [deeplearning4j/deeplearning4j#318]
Loading a 22G Google word2vec text model (or the equivalent 9G binary model) fails when using both jblas and jcublas backends. The model can be successfully loaded by Google's `distance` and `word-analogy` tools.

Code snippet:
```
    var file = new File(vecFile)
    var vec = WordVectorSerializer.loadGoogleModel(file, false)
```
Jcublas stacktrace:
```
Exception in thread "main" java.lang.IllegalArgumentException: Illegal indices [0, 0]
	at org.nd4j.linalg.api.ndarray.BaseNDArray.putScalar(BaseNDArray.java:896)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.put(BaseNDArray.java:1360)
	at org.deeplearning4j.models.embeddings.loader.WordVectorSerializer.readTextModel(WordVectorSerializer.java:112)
	at org.deeplearning4j.models.embeddings.loader.WordVectorSerializer.loadGoogleModel(WordVectorSerializer.java:77)
	at com.kt.nlp.WikiWord2Vec$.queryVec(WikiWord2Vec.scala:36)
	at com.kt.nlp.WikiWord2Vec$.main(WikiWord2Vec.scala:31)
	at com.kt.nlp.WikiWord2Vec.main(WikiWord2Vec.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
```

Jblas stacktrace:
```
Exception in thread "main" java.lang.NegativeArraySizeException
	at org.nd4j.linalg.api.buffer.BaseDataBuffer.<init>(BaseDataBuffer.java:189)
	at org.nd4j.linalg.api.buffer.FloatBuffer.<init>(FloatBuffer.java:39)
	at org.nd4j.linalg.api.buffer.factory.DefaultDataBufferFactory.createFloat(DefaultDataBufferFactory.java:42)
	at org.nd4j.linalg.factory.Nd4j.createBuffer(Nd4j.java:566)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.<init>(BaseNDArray.java:183)
	at org.nd4j.linalg.jblas.NDArray.<init>(NDArray.java:103)
	at org.nd4j.linalg.jblas.JblasNDArrayFactory.create(JblasNDArrayFactory.java:201)
	at org.nd4j.linalg.factory.BaseNDArrayFactory.create(BaseNDArrayFactory.java:193)
	at org.nd4j.linalg.factory.Nd4j.create(Nd4j.java:3109)
	at org.nd4j.linalg.factory.Nd4j.create(Nd4j.java:2500)
	at org.deeplearning4j.models.embeddings.loader.WordVectorSerializer.readTextModel(WordVectorSerializer.java:98)
	at org.deeplearning4j.models.embeddings.loader.WordVectorSerializer.loadGoogleModel(WordVectorSerializer.java:77)
	at com.kt.nlp.WikiWord2Vec$.queryVec(WikiWord2Vec.scala:36)
	at com.kt.nlp.WikiWord2Vec$.main(WikiWord2Vec.scala:31)
	at com.kt.nlp.WikiWord2Vec.main(WikiWord2Vec.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
-- org.jblas INFO Deleting /tmp/jblas6657690143696141405/libjblas.so
-- org.jblas INFO Deleting /tmp/jblas6657690143696141405/libjblas_arch_flavor.so
-- org.jblas INFO Deleting /tmp/jblas6657690143696141405
```

**Comment by agibsonccc** [comment]
My proposal for this:

https://github.com/deeplearning4j/nd4j/issues/148

I was going to prioritize speed but this storage matters a lot more.

**Comment by agibsonccc** [comment]
148 is done. Going to look closer at this tomorrow.

**Comment by agibsonccc** [comment]
@danielchalef I don't know if you're still on this or not, but I'm able to load the word2vec binary. Make sure that the last parameter is true though. (As in, don't have it load text). Please let me know how it goes with master. It'd be great to get a sanity check on your side. I'll close this in the next week or so if there's no other issues.

I still need to work on loading elements > 2^31  - 1. So the physical limit of any word2vec model with arrays will be good enough for most people. I"m going to work on making the data buffer extend to multiple arrays and buffers but for now will wait on that. It seems to load well enough for now.

**Comment by danielchalef** [comment]
Thanks! I'm out for weekend but will check early in the week.

**Comment by agibsonccc** [comment]
Found out that the google model ingestion was also borked. (It was reading the floats wrong from the input stream) 

I'd be curious to see the results now.

**Comment by danielchalef** [comment]
@agibsonccc I've given it a spin with a pull from master less than 30 minutes ago. Unfortunately I get a NegativeArraySizeException:

```
Exception in thread "main" java.lang.NegativeArraySizeException
	at org.nd4j.linalg.api.buffer.BaseDataBuffer.<init>(BaseDataBuffer.java:187)
	at org.nd4j.linalg.api.buffer.FloatBuffer.<init>(FloatBuffer.java:39)
	at org.nd4j.linalg.api.buffer.factory.DefaultDataBufferFactory.createFloat(DefaultDataBufferFactory.java:42)
	at org.nd4j.linalg.factory.Nd4j.createBuffer(Nd4j.java:566)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.<init>(BaseNDArray.java:185)
	at org.nd4j.linalg.jblas.NDArray.<init>(NDArray.java:103)
	at org.nd4j.linalg.jblas.JblasNDArrayFactory.create(JblasNDArrayFactory.java:219)
	at org.nd4j.linalg.factory.BaseNDArrayFactory.create(BaseNDArrayFactory.java:220)
	at org.nd4j.linalg.factory.Nd4j.create(Nd4j.java:3109)
	at org.nd4j.linalg.factory.Nd4j.create(Nd4j.java:2500)
	at org.deeplearning4j.models.embeddings.loader.WordVectorSerializer.readBinaryModel(WordVectorSerializer.java:145)
	at org.deeplearning4j.models.embeddings.loader.WordVectorSerializer.loadGoogleModel(WordVectorSerializer.java:68)
	at org.dc.nlp.WikiWord2Vec$.queryVec(WikiWord2Vec.scala:84)
	at org.dc.nlp.WikiWord2Vec$.main(WikiWord2Vec.scala:35)
	at org.dc.nlp.WikiWord2Vec.main(WikiWord2Vec.scala)
-- org.jblas INFO Deleting /tmp/jblas732222912044908487/libjblas.so
-- org.jblas INFO Deleting /tmp/jblas732222912044908487/libjblas_arch_flavor.so
-- org.jblas INFO Deleting /tmp/jblas732222912044908487
```

Code snippet:
```
    var file = new File(vecFile)
    var vec = WordVectorSerializer.loadGoogleModel(file, true)
```

**Comment by agibsonccc** [comment]
So I looked in to the gpu stuff a bit. The exception occurs as a disguise to the vm running oom.

**Comment by agibsonccc** [comment]
So jcublas comes from running out of heap space. I've also fixed the indexing issue there. The -array size comes from having > 2billion elements. That will require storage beyond in to support. I'm going to open a new issue that's focused on the root cause in nd4j instead.
